src/gosling-track/data-abstraction.ts

```diff
@@ -199,6 +201,68 @@ export function getTabularData(
                 }
             });
         });
+    } else if (spec.data.type === 'matrix') {
+        if (!data.dense || typeof data.tileY === 'undefined' || typeof data.tileHeight === 'undefined') {
+            // we do not have sufficient data.
+            return;
+        }
+
+        // width and height of the tile
+        const numBins = Math.sqrt(data.dense.length);
+
+        const { tileX, tileY, tileWidth, tileHeight } = data;
+        const numericValues = data.dense;
+
+        const tileXUnitSize = tileWidth / numBins;
+        const tileYUnitSize = tileHeight / numBins;
+        const columnField = spec.data.column ?? 'x';
+        const rowField = spec.data.row ?? 'y';
+
+        // TODO: a way to improve rendering performance?
+        // For the rendering performance, we aggregate cells so that we draw smaller number of marks.
+        const aggSize = 16; // assuming that # of cells can be divided by binSize, which is mostly 256 or 1024
+        for (let i = 0; i < numericValues.length / aggSize; i++) {
+            const aggLen = Math.sqrt(aggSize);
+            const xIndex = (i * aggLen) % numBins;
+            const yIndex = Math.floor((i * aggLen) / numBins) * aggLen;
+
+            // Being xIndex and yIndex the top-let origin, aggregate 4 x 4 cells
+            let value = NaN;
+            for (let c = 0; c < aggLen; c++) {
+                for (let r = 0; r < aggLen; r++) {
+                    const curVal = numericValues[(yIndex + r) * numBins + (xIndex + c)];
+                    if (!isNaN(+curVal)) {
+                        if (isNaN(value)) {
+                            value = 0;
+                        }
+                        value += curVal;
+                    }
+                }
+            }
```

Comment:
Yes, your understanding is correct and it calculates the mean. 

Actually, I think this should be summation instead of mean since cooler / matrix data calculates summation, as far as I understand correctly.

We can also support other types of aggregation functions in case a user uses matrix data with different aggregation functions.

Reply:
Removed the average calculation and added comments